crates/codegen/parser/runtime/src/query_model.rs

```diff
@@ -0,0 +1,144 @@
+use std::rc::Rc;
+
+use super::kinds::{FieldName, RuleKind, TokenKind};
+
+#[derive(Clone)]
+pub enum Query {
+    Binding(Rc<BindingMatch>),
+    Node(Rc<NodeMatch>),
+    Optional(Rc<OptionalMatch>),
+    Alternatives(Rc<AlternativesMatch>),
+    Sequence(Rc<SequenceMatch>),
+    OneOrMore(Rc<OneOrMoreMatch>),
+    Ellipsis,
+}
+
+impl Query {
+    pub fn parse(text: &str) -> Result<Self, String> {
+        super::query_parser::parse_query(text)
+    }
+
+    pub fn unparse(&self) -> String {
+        match self {
+            Self::Binding(binding) => {
+                format!("@{} {}", binding.name, binding.child.unparse())
+            }
+            Self::Node(node) => {
+                if let Some(child) = &node.child {
+                    format!("[{} {}]", node.id.unparse(), child.unparse())
+                } else {
+                    format!("[{}]", node.id.unparse())
+                }
+            }
+            Self::Optional(optional) => {
+                format!("({})?", optional.child.unparse())
+            }
+            Self::Alternatives(alternatives) => {
+                format!(
+                    "({})",
+                    alternatives
+                        .children
+                        .iter()
+                        .map(|c| c.unparse())
+                        .collect::<Vec<_>>()
+                        .join(" | ")
+                )
+            }
+            Self::Sequence(sequence) => sequence
+                .children
+                .iter()
+                .map(|c| c.unparse())
+                .collect::<Vec<_>>()
+                .join(" "),
+            Self::OneOrMore(one_or_more) => {
+                format!("({})+", one_or_more.child.unparse())
+            }
+            Self::Ellipsis => "...".to_string(),
+        }
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum Kind {
+    Rule(RuleKind),
+    Token(TokenKind),
+}
+
+impl Kind {
+    fn unparse(self) -> String {
+        match self {
+            Kind::Rule(rule) => rule.to_string(),
+            Kind::Token(token) => token.to_string(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub enum NodeId {
+    Anonymous,
+    Kind { kind: Kind },
+    String { string: String },
+    Field { field: FieldName },
+    FieldAndKind { field: FieldName, kind: Kind },
+    FieldAndString { field: FieldName, string: String },
```

Comment:
What do you think of using "text" instead of "String", which is more aligned with `Token::text`:

```suggestion
    Anonymous,
    Kind { kind: Kind },
    Text { text: String },
    Field { field: FieldName },
    FieldAndKind { field: FieldName, kind: Kind },
    FieldAndText { field: FieldName, text: String },
```

Reply:
Yes, good suggestion.